drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2129 @@
+ * @param offset Pointer to the offset from the beginning of buffer which will be incremented by the
+ * number of bytes sent to the TX FIFO
+ *
+ * @return Returns true if last byte was sent (TXLAST flag was set)
+ */
+static bool i3c_stm32_fill_tx_fifo(const struct device *dev, uint8_t *buf, size_t len,
+				   size_t *offset)
+{
+	const struct i3c_stm32_config *config = dev->config;
+	I3C_TypeDef *i3c = config->i3c;
+	bool is_last = false;
+
+	if (*offset >= len) {
+		return 0;
+	}
+
+	while (LL_I3C_IsActiveFlag_TXFNF(i3c)) {
+		if (LL_I3C_IsActiveFlag_TXLAST(i3c)) {
+			is_last = true;
+		}
+
+		if (*offset < len) {
+			LL_I3C_TransmitData8(i3c, buf[(*offset)++]);
+		}
+
+		if (is_last) {
+			return is_last;
+		}
+	}
+
+	return is_last;
+}
+
+/**
+ * @brief Drains the I3C RX FIFO from a given buffer
+ *
+ * @param buf The buffer to drain the RX FIFO to
+ * @param len The total buffer length
+ * @param offset Pointer to the offset from the beginning of buffer which will be incremented by the
+ * number of bytes drained from the RX FIFO
+ *
+ * @return Returns true if last byte was received (RXLAST flag was set)
+ */
+static bool i3c_stm32_drain_rx_fifo(const struct device *dev, uint8_t *buf, uint32_t len,
+				    size_t *offset)
+{
+	const struct i3c_stm32_config *config = dev->config;
+	I3C_TypeDef *i3c = config->i3c;
+	bool is_last = false;
+
+	if (*offset >= len) {
+		return 0;
+	}
+
+	while (LL_I3C_IsActiveFlag_RXFNE(i3c)) {
+		if (LL_I3C_IsActiveFlag_RXLAST(i3c)) {
+			is_last = true;
+		}
+
+		if (*offset < len) {
+			buf[(*offset)++] = LL_I3C_ReceiveData8(i3c);
+		}
+
+		if (is_last) {
+			return is_last;
+		}
+	}
+
+	return is_last;
+}
+
+/* Handles broadcast/direct CCCs except for ENTDAA */
+static int i3c_stm32_do_ccc(const struct device *dev, struct i3c_ccc_payload *payload)
+{
+	const struct i3c_stm32_config *config = dev->config;
+	struct i3c_stm32_data *data = dev->data;
+	I3C_TypeDef *i3c = config->i3c;
+
+	if (dev == NULL || payload == NULL) {
+		return -EINVAL;
+	}
+
+	if (payload->ccc.id == I3C_CCC_ENTDAA) {
+		return -EINVAL;
+	}
+
+	/* Check if payload has targets when sending a direct CCC */
+	if (!i3c_ccc_is_payload_broadcast(payload) &&
+	    (payload->targets.payloads == NULL || payload->targets.num_targets == 0)) {
+		return -EINVAL;
+	}
+
+	if (payload->ccc.data_len > 0 && payload->ccc.data == NULL) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->bus_mutex, K_FOREVER);
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	(void)pm_device_runtime_get(dev);
+#endif
+	/* Prevent the clocks to be stopped during the transaction */
+	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+
+	/* Mark current transfer as CCC */
+	data->msg_state = STM32_I3C_MSG_CCC;
+	data->ccc_payload = payload;
+	data->ccc_target_idx = 0;
+	data->ccc_target_payload = payload->targets.payloads;
+
+	payload->ccc.num_xfer = 0;
+
+	for (size_t i = 0; i < payload->targets.num_targets; i++) {
+		payload->targets.payloads->num_xfer = 0;
```

Comment:
this is only setting the first payload over and and over again with 0. I would expect this
```c
payload->targets.payloads[i].num_xfer = 0;
```